File: 백준/Gold/11058. 크리보드/크리보드.py

def solve(n):
    if n <= 6:
        return n

    if dp[n] != -1:
        return dp[n]

    ret = solve(n - 1) + 1

    for prv in range(1,n-2):
        ret = max(ret, solve(prv) * (n - prv - 1))

    dp[n] = ret

    return ret

if __name__ == "__main__":
    N = int(input())
    dp = [-1 for _ in range(N+1)]

    print(solve(N))

File: 백준/Gold/11058. 크리보드/크리보드.java
import java.util.*;
import java.io.*;

public class Main{
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		dp = new long[N + 1];
		Arrays.fill(dp, -1);
		System.out.println(solve(N));
	}
	static long solve(int cur) {
		if(cur <= 6 )
			return cur;
		
		if(dp[cur] != -1)
			return dp[cur];
			

		long ret = solve(cur - 1) + 1;
		for(int prv = 1; prv <= cur - 3; prv++) {
			ret = Math.max(ret,  solve(prv) * (cur - prv - 1));
		}
		
		return dp[cur] = ret;
	}
	static int N;
	static long dp[];
	
}